[P123] Add FT6206 Capacitive touch screen and [ESPEasy_TouchHandler] helper

tonhuisman opened the pull request:
Add support for FT6206 capacitive touch screen, as installed on the WT32-SC01
Resolves #3993 
Resolves #3831 

Uses an adapted version of the [Adafruit FT6206 Library](https://github.com/adafruit/Adafruit_FT6206_Library)

Functionality similar to P099 XPT2048 resistive touch screen

Introducing `ESPEasy_TouchHandler` helper module
- Define several button shapes and caption types, on/off state and enable/disable
- Define button groups to form a rudimentary menu system
- Define Slider/Gauge controls
- Handle clicking buttons
- Handle sliding a slider, with 0..100% or min/max value like -10..40 for (room)temperature
- Handle swiping, that can control the menu system
- Added an option to disable the scanning of a Touch-screen, to only use the plugin for handling on-screen buttons
- [P116] Added option for Default font, when `AdafruitGFX_Helper` has extra fonts included in the build

TODO:
- [x] Documentation
- [x] Testing
- [x] Add support for Button-like functionality, using new `ESPEasy_TouchHandler` helper
- [x] Remove [TESTING] tag before merge
- [x] Handle storing larger configuration (exceeding 1 kB)

tonhuisman committed:
[P123] Add FT6206 Capacitive touch screen

tonhuisman committed:
[AdaGFX] Implement/activate generic 'adagfx_trigger' command-trigger

tonhuisman committed:
[P123] Add arguments to Object events, allow initial events for On/Off button objects
minor corrections

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Code improvements, increased button response

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Add support for `btn` subcommand to draw button-like objects

tonhuisman committed:
[P123] Add support for button-simulation using AdaGFX btn subcommand

tonhuisman committed:
[Build] Add -Wswitch compiler option for ESP32 (already active for ESP8266)

TD-er commented on platformio_core_defs.ini:
```diff
@@ -159,7 +159,7 @@ platform_packages =
 [core_esp32_2_1_0]
 platform                  = espressif32@2.1.0
-build_flags               = -DESP32_STAGE
+build_flags               = -DESP32_STAGE -Wswitch
```
What does this build flag do?

tonhuisman replied:
Report the `switch` statements that do not implement all options of a `class enum` declaration in a `case`, while that _is_ enabled for ESP8266 (until now I used that to check my switch statements, but it would be helpful to have that for ESP32 too)

tonhuisman replied:
2 example results: (I've used a couple of these enums in  my code, so wanted to check 😉)
```
.pio/libdeps/display_ESP32_4M316k/ESPeasySerial/ESPEasySerial_ESP32.cpp: In function 'bool pinsChanged(ESPEasySerialPort, int, int)':
.pio/libdeps/display_ESP32_4M316k/ESPeasySerial/ESPEasySerial_ESP32.cpp:33:10: warning: enumeration value 'not_set' not handled in switch [-Wswitch]
   switch (port) {
          ^
```
and
```
src/src/Helpers_tmp/__tmpfile.cpp: In function 'String getPartitionType(uint8_t, uint8_t)':
src/src/Helpers_tmp/__tmpfile.cpp:7499:7: warning: case value '153' not in enumerated type 'esp_partition_subtype_t' [-Wswitch]
       case 0x99: return F("EEPROM"); // Not defined in esp_partition_subtype_t
       ^~~~
```

tonhuisman commented on platformio_esp32_envs.ini:
```diff
@@ -187,6 +187,7 @@ board                     = esp32_4M
 lib_deps                  = ${esp32_common.lib_deps}, ServoESP32
 build_flags               = ${esp32_common.build_flags}  
                             -DFEATURE_ARDUINO_OTA
+                            -DPLUGIN_DISPLAY_COLLECTION
```
This fixes the ESP32 `DISPLAY` build to again include the 'Display collection' plugins.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Minor updates and improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/P123-add-ft6206-touch-screen' of https://github.com/tonhuisman/ESPEasy-1 into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Limit code & bin size for LIMIT_BUILD_SIZE builds

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Improve btn subcommand, slim down bin size, other minor improvements

tonhuisman committed:
[TouchHandler] Introduce ESPEasy_TouchHandler, refactored out of P123 touch support

tonhuisman committed:
[P123] Refactor and implement ESPEasy_TouchHandler

tonhuisman committed:
[TouchHandler] Fix conditional compilation issue

tonhuisman committed:
[AdaGFX] Add support for confined windows to print/draw in
[AdaGFX] Add support for getting config values

tonhuisman committed:
[P095] [CP] Fix for displaying text on previous location was not overwriting old content
Add Text print mode for centered text
Extend txtfull subcommand to enable centered text

tonhuisman committed:
[TouchHandler] Move PLUGIN_WRITE and _GET_CONFIG_VALUE to helper, add getters for state and enabled, add toggle subcommand, handle multiple objects for on, off and toggle

tonhuisman committed:
[P123] Refactor PLUGIN_WRITE and _GET_CONFIG_VALUE to TouchHandler, minor improvements

tonhuisman committed:
[AdaGFX] Remove unused variable

tonhuisman committed:
[AdaGFX] Process like template for length and textheight values

tonhuisman committed:
[AdaGFX] Add/update documentation for window commands and config values

tonhuisman committed:
[AdaGFX] Code improvements, initialization, const, move window offset into printText()

TD-er commented on src/src/Helpers/AdafruitGFX_helper.cpp:
```diff
@@ -44,7 +44,7 @@
  * get the display text for a 'text print mode' enum value
  *****************************************************************************************/
-const __FlashStringHelper* toString(AdaGFXTextPrintMode mode) {
+const __FlashStringHelper* toString(const AdaGFXTextPrintMode mode) {
```
Better change it into `const xxx&` as this forces the called to have a variable of that type.
When not using a reference (or const reference), the value will be copied so any matching type can then be constructed into the type you need.
For example an int, uint or any other integer-type enum can be given.
Using a const reference requires the caller to actually have something of that type, so be prepared to see some build errors when changing it :)

tonhuisman replied:
Changed most of the `const` arguments to be passed by reference, as suggested.

TD-er commented on src/src/Helpers/AdafruitGFX_helper.cpp:
```diff
@@ -519,17 +518,17 @@ String AdaGFXparseTemplate(const String      & tmpString,
+                                       const uint16_t            fgcolor,
+                                       const uint16_t            bgcolor,
+                                       const bool                useValidation,
+                                       const bool                textBackFill)
```
idem

tonhuisman committed:
[AdaGFX] Code improvements as suggested by feedback

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Correct font include paths, add #ifndef filename checks to fonts

tonhuisman committed:
[AdaGFX] Regression: Background-fill for text should fill entire line height

tonhuisman committed:
[TouchHandler] Change method arguments to const-by-reference where possible

tonhuisman committed:
[TouchHandler] Strip unneeded prefix from page variables

tonhuisman committed:
[P123] Strip unneeded prefix from page variables

TD-er commented on src/src/Helpers/ESPEasy_TouchHandler.cpp:
```diff
@@ -1181,32 +1181,32 @@ bool ESPEasy_TouchHandler::plugin_webform_save(struct EventStruct *event) {
-  bitWrite(lSettings, TOUCH_FLAGS_INIT_OBJECTEVENT, isFormItemChecked(F("tch_init_objectevent")));
+  bitWrite(lSettings, TOUCH_FLAGS_SEND_XY,          bitRead(getFormItemInt(F("events")), TOUCH_FLAGS_SEND_XY));
+  bitWrite(lSettings, TOUCH_FLAGS_SEND_Z,           bitRead(getFormItemInt(F("events")), TOUCH_FLAGS_SEND_Z));
+  bitWrite(lSettings, TOUCH_FLAGS_SEND_OBJECTNAME,  bitRead(getFormItemInt(F("events")), TOUCH_FLAGS_SEND_OBJECTNAME));
```
Beter store the int in a const variable before processing, as it now has to be found several times in the args

tonhuisman committed:
[TouchHandler] Optimizations (as suggested and more)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

TD-er commented on src/_P123_FT62x6Touch.ino:
```diff
@@ -122,19 +123,19 @@ boolean Plugin_123(uint8_t function, struct EventStruct *event, String& string)
+      AdaGFXFormColorDepth(F("colordepth"), P123_COLOR_DEPTH, (colorDepth_ == 0));
 
-      addFormNumericBox(F("Touch minimum pressure"), F("p123_treshold"), P123_CONFIG_TRESHOLD, 0, 255);
+      addFormNumericBox(F("Touch minimum pressure"), F("treshold"), P123_CONFIG_TRESHOLD, 0, 255);
```
I know this is less characters, but my OCD is missing a letter here ;)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Fix some spelling issues, code optimizations

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Minor optimizations

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Reduce build some more for LIMIT_BUILD_SIZE

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Small size improvement

tonhuisman committed:
[TouchHandler] Replace _ in captions and object names by a space, and reverse, might save some settings size

tonhuisman committed:
[UI] Add class table.multi2row for alternating color per 2 rows

tonhuisman committed:
[TouchHandler] Use 2-row alternating color, improvements

tonhuisman committed:
[P123] Code improvements

tonhuisman commented:
An example of the alternating color per 2 rows css class (`"multi2row"`): (Updated 2022-08-29 after changed css is merged, and a few adjustments applied)

![Screenshot - 29-08-2022 , 23_19_59](https://user-images.githubusercontent.com/1922393/187301097-d9d549a2-a16f-4cd3-962d-f88e07c811fa.png)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Add Slider/Gauge support (via btn subcommand), fixes

tonhuisman committed:
[AdaGFX] Missed a conditional compile option

tonhuisman committed:
[TouchHandler] Add support for sliders and swiping, many other improvements

tonhuisman committed:
[P123] Implement Slider/Gauge support, other (size) improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Improved drawing of Slider/Gauge control (also reverted)

tonhuisman committed:
[TouchHandler] Enable use of reverted sliders/gauges, and improved handling of ranges that don't have 0 included

tonhuisman committed:
[P123] Minor code improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler] Add optional touchscreen disabling, code fixes and improvements

tonhuisman committed:
[P123] Implement disabled-touchscreen feature

tonhuisman committed:
[TouchHandler] Fix typo and conditional compilation issue

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman commented:
Hmm, merging css/minified css/WebStaticData.h is quite a painful operation 😞

tonhuisman committed:
[UI] Improvements for `multi2row` style

tonhuisman committed:
[TouchHandler] Settings page layout adjusted, and some minor improvements

tonhuisman committed:
[TouchHandler] Logging check improvements

tonhuisman committed:
[P123] Logging check improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman commented:
I reverted my css modifications (table highlight per 2 rows) for now, and will redo that later once all pending css changes are settled, the merging causes me too much of a headache 🤔

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/P123-add-ft6206-touch-screen' of https://github.com/tonhuisman/ESPEasy-1 into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler] Remove unneeded enum MAX values

tonhuisman committed:
[AdaGFX] Remove unneeded enum MAX values, fix BMP conditional compilation issue

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX] Fix compilation error after merge from mega

tonhuisman committed:
[TouchHandler] Fix compilation error after merge from mega, table improvements

tonhuisman committed:
[P116] Add GetConfig feature, code improvements

tonhuisman committed:
[I2C scanner] Enable plugin-scan for ESP32, updated sensors

tonhuisman committed:
[TouchHandler] Fix touch-disabled option, code improvements

tonhuisman committed:
[P123] Code improvements

tonhuisman committed:
[TouchHandler] Conditional compilation fix

tonhuisman committed:
[TouchHandler] Conditional compilation fix

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[Boards] Add BOARD_HAS_PSRAM flag for 16M8M config (avoids crashing when saving P123 settings)

tonhuisman committed:
[TouchHandler] Logging and code optimizations

tonhuisman committed:
[P123] Bugfixes and logging improvements

tonhuisman committed:
[P123] Fix for disabled touchpanel setting

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[Lib FT6206] Add support for FT6336 touch panel (as used in M5Stack Core2)

TD-er commented on platformio_core_defs.ini:
```diff
@@ -245,10 +245,10 @@ build_flags                 = -DESP32_STAGE
-platform                    = https://github.com/Jason2866/platform-espressif32.git
-platform_packages           = framework-arduinoespressif32 @ https://github.com/tasmota/arduino-esp32/releases/download/2.0.1/framework-arduinoespressif32-release_IDF4.4.tar.gz
-                              platformio/tool-esptoolpy @ https://github.com/tasmota/esptool/releases/download/v3.2/esptool-v3.2.zip
+platform                    = https://github.com/tasmota/platform-espressif32/releases/download/v2.0.4.1/platform-espressif32-2.0.4.1.zip
```
Why 2.0.4.1 and not 2.0.5.1?

tonhuisman replied:
Hmm, must have mixed up a merge conflict 😞

tonhuisman replied:
But AFAICS, that definition isn't currently used 😅

TD-er replied:
You sure? ;)
https://github.com/letscontrolit/ESPEasy/blob/86cd66816f538b4505f6d3f38734c18d3db52251/platformio_core_defs.ini#L236

tonhuisman replied:
The merge error is in the `[core_esp32_stage]` section, in the `[core_esp32_IDF4_4__2_0_5]` section that is referenced from current `env:` sections the correct version is used.

tonhuisman committed:
[P123] Correct merge error

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] De-initialize on error during PLUGIN_INIT fixed

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/P123-add-ft6206-touch-screen' of https://github.com/tonhuisman/ESPEasy-1 into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/P123-add-ft6206-touch-screen' of https://github.com/tonhuisman/ESPEasy-1 into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/Core-extend-custom-task-settings' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/Converters-add-tostringnozero' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler][P123] Implement Extended CustomTaskSettings feature

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/Core-extend-custom-task-settings' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/Core-extend-custom-task-settings' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123][TouchHandler] Fix merge issues, adjust log strings

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

TD-er commented on src/src/Helpers/ESPEasy_TouchHandler.cpp:
```diff
@@ -1685,18 +1665,11 @@ bool ESPEasy_TouchHandler::plugin_fifty_per_second(struct EventStruct *event,
       Touch_Settings.logEnabled &&
 
-      // REQUIRED for calibration and setting up objects, so do not make this optional!
+      // Thisi log is REQUIRED for calibration and setting up objects, so do not make this optional!
```
typo :)

TD-er commented on src/src/Helpers/ESPEasy_TouchHandler.cpp:
```diff
@@ -1706,14 +1679,26 @@ bool ESPEasy_TouchHandler::plugin_fifty_per_second(struct EventStruct *event,
         Device[DeviceIndex].VType      = Sensor_VType::SENSOR_TYPE_DUAL;
         Device[DeviceIndex].ValueCount = 2;
+        # else // ifdef ESP8266
+        Device.getDeviceStructForEdit(DeviceIndex).VType      = Sensor_VType::SENSOR_TYPE_DUAL;
```
That function was there as a reminder not to use it....
Device vector should be considered const.

tonhuisman replied:
Yup, understood, but there currently isn't a feature available to have variation in the number of values to send based on task settings, AFAICS, so this is the least worse solution.
NB: Copied that from P099 😋

TD-er replied:
Ah check, so that's up to me to do some work ;)

tonhuisman replied:
Or it could be implemented using the `PLUGIN_GET_DEVICEVTYPE` plugin event. Currently it's changed 'on the fly', just before sending the data to the controller, but that shouldn't be needed once set via the event 🤔.

TD-er replied:
Problem with non-const Devices array is that such changes aren't task specific.

tonhuisman replied:
I'm currently switching to `PLUGIN_GET_DEVICEVALUECOUNT`, as that is actually the _correct_ place to set the number of output values for this plugin. No need to change the `Device_VType` setting, so that was in fact a bug 😞

tonhuisman committed:
[TouchHandler] Fix typo

tonhuisman committed:
[P123][TouchHandler] Fix valuecount issue, implement missing `GET_I2C_ADDRESS` function

tonhuisman committed:
[TouchHandler] Remove unused variable assignment

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler] Fix compilation errors since Core code changes

tonhuisman committed:
[TouchHandler] Fix compilation errors since Core code changes

tonhuisman committed:
[TouchHandler] Code optimization reducing .bin size

tonhuisman committed:
[P123] Minor code optimiziations

tonhuisman committed:
[Core] String optimizations

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'feature/P123-add-ft6206-touch-screen' of https://github.com/tonhuisman/ESPEasy-1 into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler] Also reverse the Left/Right navigation when Up/Down navigation is reversed

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[TouchHandler] Rename inc/dec subcommands to next/prev

tonhuisman committed:
[P123] Refactor increment/decrement functions to next/prev in line with TouchHandler

tonhuisman committed:
[P123][TouchHandler] Add documentation

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[P123] Docs minor corrections

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[Storage] Use correct parameter-type for strformat()

tonhuisman committed:
[Storage] Use correct parameter-type for strformat()

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX_Helper] Fix typo in `default` font definition

tonhuisman committed:
[P116] Add Default font selection if AdafruitGFX_Helper fonts are included in build

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
[AdaGFX_Helper] Fix merge conflict

tonhuisman committed:
[AdaGFX_Helper] Fix font offset issues

saint-hh commented:
Just to resume:
I can test this latest build regarding the display and fix my rules for it.
The SEN5X is not supported in this build, but all will come together in the next official release. Is this correct?
(just to make myself a "workflow")

tonhuisman commented:
> Just to resume: I can test this latest build regarding the display and fix my rules for it. The SEN5X is not supported in this build, but all will come together in the next official release. Is this correct? (just to make myself a "workflow")

~Yes, the 'old'/original version of the SEN5x plugin, that only supports the IKEA Vindstyrka, is included, as I pulled in the `mega` branch. Don't think that will work with the standalone SEN5x sensors.~ Plugin is not yet included in any build 😞 but that's taken care of in the update PR for P167.

It will all come together once this plugin is merged, hope to have it tested enough soon, as this PR has been started > 2 years ago... 😲

tonhuisman commented:
> the 'old'/original version of the SEN5x plugin, that only supports the IKEA Vindstyrka, is included,

Ah, snap, it's _not_ included, as that was one of the issues with that PR, the plugin wasn't included in any build... 😮

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P123-add-ft6206-touch-screen

tonhuisman committed:
Merge branch 'mega' into feature/P123-add-ft6206-touch-screen
